### PR TITLE
migration: Add case to test disk size

### PR DESCRIPTION
**Pull request by cliping**
XXX-300490 - VM live migration with copy storage - migrate a disk to an equal or larger disk size

### CODE REVIEW

**Comment by cliping** [comment]
Depend on: https://github.com/avocado-framework/avocado-vt/pull/3889

**Comment by cliping** [comment]
Test result:

 (1/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_raw.src_disk_raw.non_p2p: STARTED
 (1/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_raw.src_disk_raw.non_p2p: PASS (226.76 s)
 (2/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_raw.src_disk_block.non_p2p: STARTED
 (2/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_raw.src_disk_block.non_p2p: PASS (536.98 s)
 (3/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_block.src_disk_raw.non_p2p: STARTED
 (3/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_block.src_disk_raw.non_p2p: PASS (571.74 s)
 (4/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_block.src_disk_block.non_p2p: STARTED
 (4/4) type_specific.io-github-autotest-libvirt.migration_with_copy_storage.target_image_larger_than_source.target_disk_block.src_disk_block.non_p2p: PASS (729.14 s)

**Review by nanli1 — APPROVED** [review]

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            if not lv_utils.vg_check(src_vg_name):
+                lv_utils.vg_create(src_vg_name, dev_src)
+            if not lv_utils.lv_check(src_vg_name, src_lv_name):
+                lv_utils.lv_create(src_vg_name, src_lv_name, disk_size)
+            cmd = "qemu-img convert -f qcow2 -O %s %s %s" % (disk_format, blk_source, src_block_path)
+        else:
+            cmd = "qemu-img convert -f qcow2 -O %s %s %s" % (disk_format, blk_source, disk_source_name)
+        process.run(cmd, shell=True)
+
+        libvirt.set_vm_disk(vm, params)
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:74]: Line 69-72 seems included in libvirt.set_vm_disk().  So I feel a little duplicated codes here. Could you help check again?

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            libvirt_disk.create_disk("file", path=disk_source_name,
+                                     size=target_disk_size, disk_format=disk_format,
+                                     session=remote_session)
+            remote_session.close()
+
+        outxml = virsh.dumpxml(vm_name, "--migratable", debug=True).stdout_text.strip()
+        mig_disk = disk.Disk(type_name=target_disk_type)
+        mig_disk.setup_attrs(**target_disk_dict)
+
+        tmp_file = tempfile.mktemp(dir=data_dir.get_tmp_dir())
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:97]: This function tempfile.mktemp is deprecated. Please use alternative method

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        mig_xml = ElementTree.fromstring(outxml)
+        dev = mig_xml.find("./devices")
+        dev_disk = mig_xml.find(".//disk")
+        dev.remove(dev_disk)
+        new_disk = ElementTree.parse(mig_disk.xml).getroot()
+        dev.append(new_disk)
+        ElementTree.ElementTree(mig_xml).write(tmp_file)
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:104]: For line 98 to 104, ElementTree is too low level API and we do not recommend to directly use it here. Could you find other upper layer API as a better implementation?

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        target_disk_dict = eval(params.get("target_disk_dict"))
+        src_iscsi_size = params.get("src_iscsi_size")
+        target_iscsi_size = params.get("target_iscsi_size")
+
+        test.log.debug("Setup steps.")
+        migration_obj.conn_list.append(migration_base.setup_conn_obj(migrate_desturi_type, params, test))
+        migration_obj.remote_add_or_remove_port(migrate_desturi_port)
+
+        blk_source = vm.get_first_disk_devices()['source']
+        if disk_type == "block":
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:62]: Too many lines are in this function. For better readibility, could you help split the steps for preparing image to other functions?

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        """
+        Verify steps
+
+        """
+        test.log.debug("Verify steps.")
+        virsh.blockresize(vm_name, "vda", target_disk_size, ignore_status=False, uri=desturi)
+        migration_obj.verify_default()
+
+    def cleanup_test():
+        """
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:117]: Could you help split below into two methods, one is to remove for source, the other is to remove for target.

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py (hunk 1 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py]: One of scenarios in this case need to create two block disks, so we can't use libvirt.set_vm_disk() directly.

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py (hunk 2 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py (hunk 3 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py (hunk 4 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py (hunk 5 above)
Comment (cliping) [reply libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:244]: Updated, thanks.

File: libvirt/tests/cfg/migration_with_copy_storage/target_image_larger_than_source.cfg [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    ssh_timeout = 60
+    # Local URI
+    virsh_migrate_connect_uri = "qemu:///system"
+    image_convert = "no"
+    migrate_desturi_port = "16509"
+    migrate_desturi_type = "tcp"
+    virsh_migrate_desturi = "qemu+tcp://${migrate_dest_host}/system"
+    setup_nfs = "no"
+    server_ip = "${migrate_dest_host}"
+    server_user = "root"
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/cfg/migration_with_copy_storage/target_image_larger_than_source.cfg]: Updated, thanks.

File: libvirt/tests/cfg/migration_with_copy_storage/target_image_larger_than_source.cfg [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    server_pwd = "${migrate_dest_pwd}"
+    client_ip = "${migrate_source_host}"
+    client_user = "root"
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/cfg/migration_with_copy_storage/target_image_larger_than_source.cfg]: Updated, thanks.

File: libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+            lv_utils.vg_create(src_vg_name, dev_src)
+        if not lv_utils.lv_check(src_vg_name, src_lv_name):
+            lv_utils.lv_create(src_vg_name, src_lv_name, disk_size)
+        cmd = "qemu-img convert -f qcow2 -O %s %s %s" % (disk_format, blk_source, src_block_path)
+    else:
+        cmd = "qemu-img convert -f qcow2 -O %s %s %s" % (disk_format, blk_source, disk_source_name)
+    process.run(cmd, shell=True)
+
+
+def Setup_migratable_xml(vm_name, params):
```
Comment (chunfuwen) [inline libvirt/tests/src/migration_with_copy_storage/target_image_larger_than_source.py:69]: use capital letter for this function, should setup_migratable_xml?

**Review by dzhengfy — APPROVED** [review]
LGTM

**Review by chunfuwen — APPROVED** [review]
lgtm

**Pull request merged**